tools/autograd/templates/Functions.cpp

```diff
@@ -1961,6 +1961,38 @@ Tensor log1p_backward(const Tensor& grad, const Tensor& self) {
   return grad / (self + 1);
 }
 
+Tensor constant_pad_nd_backward(const Tensor& grad, const Tensor& self, IntList pad) {
+    auto l_inp = self.dim();
+    auto l_pad = pad.size() / 2;
+    auto l_diff = l_inp - l_pad;
+
+    auto grad_input = Variable(at::zeros(self.sizes(), grad.options()));
+
+    auto cg_input = grad_input;
+    for (int i = l_diff; i < l_inp; i++) {
+        auto pad_idx = pad.size() - (i - l_diff + 1) * 2;
+        if (pad[pad_idx] < 0) {
+            cg_input = cg_input.narrow(i, -pad[pad_idx], cg_input.size(i) + pad[pad_idx]);
+        }
+        if (pad[pad_idx + 1] < 0) {
+            cg_input = cg_input.narrow(i, 0, cg_input.size(i) + pad[pad_idx + 1]);
+        }
+    }
+
+    auto cg_output = grad;
+    for (int i = l_diff; i < l_inp; i++) {
+        auto pad_idx = pad.size() - (i - l_diff + 1) * 2;
+        if (pad[pad_idx] > 0) {
+            cg_output = cg_output.narrow(i, pad[pad_idx], cg_output.size(i) - pad[pad_idx]);
+        }
+        if (pad[pad_idx + 1] > 0) {
+            cg_output = cg_output.narrow(i, 0, cg_output.size(i) - pad[pad_idx + 1]);
+        }
+    }
+    cg_input.copy_(cg_output);
```

Comment:
Actually I think just returning the sliced `cg_output` should work. It will be much more efficient as well.

Reply:
I tested it out, and I'm pretty sure there are problems with the shape of the output if you just return the sliced `cg_output` and there are negative pads involved. I made a notebook of the case I ran through (https://gist.github.com/wdhorton/9fa988750bab88046168a81a813e468f), but it's basically this case:
```
input of size (4,5)
pad is (1, 1, -1, -1)
grad of size (2, 7)

If you return cg_output in this case, the size is (2, 5).
But you want it to be (4, 5), matching the original input.
```

Let me know if that makes sense, or if I'm missing anything.

Reply:
Thinking about it more, we could have some logic conditional on whether there are only positive pads to avoid the `copy_`, and that would be an improvement for that case (which is probably more common anyway)

Reply:
Ah I see. Thanks for the explanation. Yes, it would be great if we can have some logic to detect if only positive padding amounts exist. It would save creating a new tensor and copying.

Actually, here is a proposal:
the backward of `pad(input, pads)` is just `pad(grad_output, [-p for p in pads])`. So we can also specialize all negative paddings in **forward** mode and just write the backward entry as a three-liner:
```cpp
#include <algorithm>
#include <functional>
.....
{
  auto negated_pad = pad.vec();
  std::transform(negated_pad.cbegin(), negated_pad.cend(), Mynegated_padec.begin(), std::negate<int64_t>());
  return constant_pad_nd_backward(grad, negated_pad);
}
```
You can avoid saving `self` too :)